addOns/exim/src/main/java/org/zaproxy/addon/exim/pcap/PcapUtils.java

```diff
@@ -0,0 +1,234 @@
+        for (TCPPacket tcpPacket : tcpPackets) {
+            if (tcpPacket.getPayload() != null
+                    && requestDirection.equals(new TransportStreamId(tcpPacket))) {
+                requestFlow.append(
+                        tcpPacket
+                                .getPayload()
+                                .toString()); // naively reassemble the client requests
+            }
+        }
+        return requestFlow.toString();
+    }
+
+    public static byte[] getHttpResponseFlow(TcpStream stream) {
+        List<TCPPacket> tcpPackets = stream.getPackets();
+        ByteArrayOutputStream responseFlow = new ByteArrayOutputStream();
+        TransportStreamId responseDirection =
+                new TransportStreamId(tcpPackets.get(0))
+                        .oppositeFlowDirection(); // server is receiving connection
+        for (TCPPacket tcpPacket : tcpPackets) {
+            if (tcpPacket.getPayload() != null
+                    && responseDirection.equals(new TransportStreamId(tcpPacket))) {
+                try {
+                    responseFlow.write(tcpPacket.getPayload().getArray());
+                } catch (IOException e) {
+                    LOGGER.error("Failed to write response payload: " + e.getMessage());
+                }
+            }
+        }
+        return responseFlow.toByteArray();
+    }
+
+    public static List<HttpMessage> constructHttpMessages(String requestFlow, byte[] responseFlow) {
+        List<HttpMessage> requestsMessages = constructHttpRequests(requestFlow);
+        return constructHttpResponses(responseFlow, requestsMessages);
+    }
+
+    private static List<HttpMessage> constructHttpRequests(String requestFlow) {
+        List<HttpMessage> requests = new ArrayList<>();
+        ArrayList<String> flow = new ArrayList<>(List.of(requestFlow.split(DOUBLE_CRLF)));
+
+        ListIterator<String> flowIt = flow.listIterator();
+
+        while (flowIt.hasNext()) {
+            HttpRequestHeader reqHeader = new HttpRequestHeader();
+            HttpRequestBody reqBody = new HttpRequestBody();
+            try {
+                reqHeader = new HttpRequestHeader(flowIt.next());
+                if (flowIt.hasNext()) {
+                    try {
+                        // if the lookahead is a header, then the request has an empty body
+                        HttpRequestHeader lookahead = new HttpRequestHeader(flowIt.next());
+                        flowIt.previous();
+                    } catch (HttpMalformedHeaderException e) {
+                        flowIt.previous();
+                        reqBody = new HttpRequestBody(flowIt.next());
+                    }
+                }
+            } catch (HttpMalformedHeaderException e) {
+                // we assume the header will always come first
+                LOGGER.error("Failed to parse request header: " + e.getMessage());
```

Comment:
- Use `{}`.
- Error is usually reserved for actual errors/bugs in ZAP, here I'd go with warn since it's bad input.
(Same comments for the other log message later)